src/scripts/client/modules/hw.js

```diff
@@ -0,0 +1,200 @@
+
+
+  function addMainContent( $ui, resp ) {
+    $ui
+      .html( "" )
+      .append( `<h5>Newline HW ${resp.status === "ok" ? 'detected' : 'not found'}</h5>` );
+
+    if ( resp.status === "ok" ) {
+      $ui.append( `Last Heartbeat at ${moment(resp.message_at).fromNow()}` );
+      $ui.append( `<dl>` );
+
+      for ( let i in resp.data ) {
+        if ( resp.data.hasOwnProperty( i ) ) {
+          $ui.append( `<dt>${i}</dt><dd>${resp.data[i]}</dd>` );
+        }
+      }
+
+      $ui.append( `</dl>` );
+    } else {
+      $.get( chrome.extension.getURL( HW_NO_CLIENT_TEMPLATE ) ).then( function( html ) {
+        $ui.append( html );
+      } );
+    }
+  }
+
+
+  function addCloneLinksToUi() {
+    detectNativeClient( function handleNativeClient( resp ) {
+      if ( resp.status === "ok" ) {
+        if ( pageData.assignment_submission ) {
+          updateUIOnSubmissionPage();
+        }
+        if ( pageData.assignment ) {
+          updateUIOnAssignmentPage();
+        }
+      }
+    } );
+  }
+
+
+  function updateUIOnAssignmentPage() {
+    const idFromUrl = uri => Number( uri.substr( uri.lastIndexOf( '/' ) + 1 ) );
+    $( "#submissions a[href^='/admin/assignment_submissions']" ).closest( "td" ).each( function addLinks() {
+      const $el = $( this );
+      addCloneLinkForSubmissionTo(
+        $el,
+        idFromUrl( $el.find( "a[href^='/admin/assignment_submissions']" )[ 0 ].href ), {
+          success_class: "label label-complete-and-satisfactory",
+          fail_class: "label label-external-link"
+        }
+      );
+    } );
+  }
+
+
+  function updateUIOnSubmissionPage() {
+    addCloneLinkForSubmissionTo(
+      $( ".edit_assignment_submission" ).first(),
+      pageData.assignment_submission.id, {
+        success_class: "btn btn-primary btn-sm",
+        fail_class: "btn btn-info btn-sm"
+      }
+    );
+  }
+
+  /**
+   * Adds a link to an $el that will trigger a triggerCloneEventForSubmissionID
+   * event for a submission_id when clicked.  If a submission_id cannot be cloned
+   * display a revelant message to the user.
+   *
+   * @param {jQuery Element} $el the element to append the relevant link
+   * to
+   * @param {integer} submission_id the submission_id that is to be passed to
+   * NewlineHW
+   * @param {object} options configuration options for the link (currently
+   * on success class and fail class)
+   */
+  function addCloneLinkForSubmissionTo( $el, submission_id, options = {} ) {
+    isSubmissionCloneable( submission_id, function handleIsSubmissionCloneable( resp ) {
+      if ( resp.data.cloneable ) {
+        const clone_link = $el.append(
+          `
+          <a class="clone_and_open_submission ${options.success_class}">
+            <i class="fa fa-download" aria-hidden="true"></i>
+            Clone Assignment Locally
+          </a>
+          `
+        );
+
+        clone_link
+          .find( '.clone_and_open_submission' )
+          .click( function getSubmissionIDandCall( e ) {
```

Comment:
It would be great to make this a delegated event handler instead (more efficient). Not a blocker, but let me know if you want to try it.